test/e2e/receive_test.go

```diff
@@ -8,222 +8,227 @@ import (
+			Deduplicate: false,
+		}, []model.Metric{
+			{
+				"job":        "myself",
+				"prometheus": "prom1",
+				"receive":    "2",
+				"replica":    "0",
+			},
+			{
+				"job":        "myself",
+				"prometheus": "prom2",
+				"receive":    "1",
+				"replica":    "0",
+			},
+			{
+				"job":        "myself",
+				"prometheus": "prom3",
+				"receive":    "3",
+				"replica":    "0",
+			},
+		})
+	})
+
+	t.Run("replication", func(t *testing.T) {
+		t.Parallel()
+
+		s, err := e2e.NewScenario("e2e_test_receive_replication")
+		testutil.Ok(t, err)
+		defer s.Close()
+
 		// The replication suite creates three receivers but only one
 		// receives Prometheus remote-written data. The querier queries all
 		// receivers and the test verifies that the time series are
 		// replicated to all of the nodes.
-		receiveGRPC1, receiveGRPC2, receiveGRPC3 := a.New(), a.New(), a.New()
+		r1, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "1", 3)
+		testutil.Ok(t, err)
+		r2, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "2", 3)
+		testutil.Ok(t, err)
+		r3, err := e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "3", 3)
+		testutil.Ok(t, err)
 
 		h := receive.HashringConfig{
-			Endpoints: []string{receiveGRPC1.HostPort(), receiveGRPC2.HostPort(), receiveGRPC3.HostPort()},
+			Endpoints: []string{
+				r1.GRPCNetworkEndpointFor(s.NetworkName()),
+				r2.GRPCNetworkEndpointFor(s.NetworkName()),
+				r3.GRPCNetworkEndpointFor(s.NetworkName()),
+			},
 		}
 
-		r1 := receiver(a.New(), receiveGRPC1, a.New(), 3, h)
-		r2 := receiver(a.New(), receiveGRPC2, a.New(), 3, h)
-		r3 := receiver(a.New(), receiveGRPC3, a.New(), 3, h)
-
-		prom1 := prometheus(a.New(), defaultPromConfig("prom1", 1, remoteWriteEndpoint(r1.HTTP)))
-
-		q1 := querier(a.New(), a.New(), []address{r1.GRPC, r2.GRPC, r3.GRPC}, nil)
-
-		testCases = append(testCases, receiveTestConfig{
-			name:         "replication",
-			cmds:         []scheduler{q1, prom1, r1, r2, r3},
-			queryAddress: q1.HTTP,
-			metrics: []model.Metric{
-				{
-					"job":        "test",
-					"prometheus": "prom1",
-					"receive":    model.LabelValue(r1.HTTP.Port),
-					"replica":    "1",
-				},
-				{
-					"job":        "test",
-					"prometheus": "prom1",
-					"receive":    model.LabelValue(r2.HTTP.Port),
-					"replica":    "1",
-				},
-				{
-					"job":        "test",
-					"prometheus": "prom1",
-					"receive":    model.LabelValue(r3.HTTP.Port),
-					"replica":    "1",
-				},
-			}})
-	}
-	{
-		// The replication suite creates a three-node hashring but one of the
-		// receivers is dead. In this case, replication should still
-		// succeed and the time series should be replicated to the other nodes.
-		receiveGRPC1, receiveGRPC2, receiveGRPC3 := a.New(), a.New(), a.New()
+		// Recreate again, but with hashring config.
+		r1, err = e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "1", 3, h)
```

Comment:
Shouldn't we close the previous ones before trying to start these up? AFAICT it does an equivalent of `docker run ...` later on and it doesn't stop the previously created ones.

Reply:
`s.Close` does it.